src/GovUk.Education.ExploreEducationStatistics.Common.Tests/Extensions/DbContextTransactionExtensionsTests.cs

```diff
@@ -0,0 +1,364 @@
+                        return new Either<int, string>(1);
+                    });
+                });
+
+                AssertNoEntitiesInAnyDbContexts();
+            }
+
+            [Fact]
+            public async Task FailAtTopLevelWithNestedTransaction()
+            {
+                await Assert.ThrowsAsync<SimulateFailureException>(() => GetDbContext1()
+                    .RequireTransaction(async () =>
+                    {
+                        await WriteToAllDbContexts();
+                        await GetDbContext1().RequireTransaction(() => WriteToAllDbContexts(2));
+                        throw new SimulateFailureException();
+                    }));
+
+                AssertNoEntitiesInAnyDbContexts();
+            }
+
+            [Fact]
+            public async Task FailAtTopLevelWithNestedTransaction_WithEither()
+            {
+                await GetDbContext1().RequireTransaction(async () =>
+                {
+                    await WriteToAllDbContexts();
+                    await GetDbContext1().RequireTransaction(() => WriteToAllDbContexts(2));
+                    return new Either<int, string>(1);
+                });
+
+                AssertNoEntitiesInAnyDbContexts();
+            }
+        }
+    }
+
+    private void AssertEntitiesInAllDbContexts(int expectedId = 1)
+    {
+        Assert.NotNull(GetDbContext1()
+            .Entities
+            .SingleOrDefaultAsync(entity => entity.Id == expectedId));
+
+        Assert.NotNull(GetDbContext2()
+            .Entities
+            .SingleOrDefaultAsync(entity => entity.Id == expectedId));
+
+        Assert.NotNull(GetDbContext3WithoutRetry()
+            .Entities
+            .SingleOrDefaultAsync(entity => entity.Id == expectedId));
+    }
+
+    private void AssertNoEntitiesInAnyDbContexts()
+    {
+        Assert.Empty(GetDbContext1().Entities);
+        Assert.Empty(GetDbContext2().Entities);
+        Assert.Empty(GetDbContext3WithoutRetry().Entities);
+    }
+
+    internal class TestEntity
+    {
+        public int Id { get; set; }
+    }
+
+    internal class TestDbContext1(
+        DbContextOptions<TestDbContext1> options) : DbContext(options)
+    {
+        public DbSet<TestEntity> Entities { get; init; } = null!;
+    }
+
+    internal class TestDbContext2(
+        DbContextOptions<TestDbContext2> options) : DbContext(options)
+    {
+        public DbSet<TestEntity> Entities { get; init; } = null!;
+    }
+
+    internal class TestDbContext3WithoutRetry(
+        DbContextOptions<TestDbContext3WithoutRetry> options) : DbContext(options)
+    {
+        public DbSet<TestEntity> Entities { get; init; } = null!;
+    }
+
+    private class SimulateFailureException : Exception;
+
+    // ReSharper disable once ClassNeverInstantiated.Global
+    public class DbContextTransactionExtensionsTestFixture :
+        IAsyncLifetime
+    {
+        public readonly PostgreSqlContainer[] PostgreSqlContainers =
```

Comment:
I suspect it would, but I think I would err on leaving that til a different PR if that's OK.  Given we don't have a Test Container for SQL Server just yet, we might have to rely on manual testing of this initially.

Reply:
Yeah that's cool with me! We technically test this as part of some of the endpoints anyway